### PR TITLE
Build and test improvements

### PR DESCRIPTION
**Pull request by costinm**
- optimize and make explicit some of docker builds - use pre-build base debug image (will be generated by nightly script )
- use separate targets for large binaries, to avoid one directory with lots of binaries (slows down builds)
- add make target optimized for local run in minikube (VM or root) and developer flow.
- the local targets use specified namespaces, can be run in parallel on the same cluster !
- move and make consistent the downloading of logs (using the one from e2e_ in pilot)
- apply the 'v1alpha' flag for egress and other v1alpha1 were it was missing
- explicit targets for auth and noauth

Also added an option so developers can build images based on the 'debug' image, with curl and debug
tools included. Currently applies to the proxy image ( using release/no-debug envoy ) and the pilot test app, if other people find this option useful we can expand to pilot, mixer and all other scratch-based images. This does not affect the release build - and is off by default, add to .istiorc.mk if you find it useful.

Also, by popular demand, the 'docker' target will also build. Circleci will still use the split mode, using
'docker.all' target which just packs all docker images.

### CODE REVIEW

File: pilot/docker/Dockerfile.proxy_debug [hunk 1]
```diff
@@ -1,40 +1,15 @@
...
-      netcat \
-      tcpdump \
-      net-tools \
-      libc6-dbg gdb \
-      elvis-tiny \
-      lsof \
-      busybox \
-      sudo && \
-    rm -rf /var/lib/apt/lists/*
+FROM istionightly/base_debug
```
Comment (ldemailly) [inline pilot/docker/Dockerfile.proxy_debug]: let's not base our releases on your org and unknown content
please 
- commit the dockerfile for that base
- make that base image exist in istio/ first

File: pilot/docker/Dockerfile.proxy_debug [hunk 2]
```diff
@@ -1,40 +1,15 @@
...
 
 COPY envoy_pilot.json      /etc/istio/proxy/envoy_pilot.json
 COPY envoy_pilot_auth.json /etc/istio/proxy/envoy_pilot_auth.json
 COPY envoy_mixer.json      /etc/istio/proxy/envoy_mixer.json
 COPY envoy_mixer_auth.json /etc/istio/proxy/envoy_mixer_auth.json
-# Use the same file as the deb file.
 COPY envoy_bootstrap_tmpl.json /var/lib/istio/envoy/envoy_bootstrap_tmpl.json
 
-# Sudoers used to allow tcpdump and other debug utilities.
-RUN useradd -m --uid 1337 istio-proxy && \
```
Comment (ldemailly) [inline pilot/docker/Dockerfile.proxy_debug]: why are you removing this / where is the replacement for it ?

File: pilot/docker/Dockerfile.proxy_debug (hunk 1 above)
Comment (nmittler) [inline pilot/docker/Dockerfile.proxy_debug]: @costinm why are we re-introducing a dependency on another Dockerfile? This was the source of a lot of confusion previously.  If we wanted to try out local changes to the base setup, it would seem that relying on an external image would be troublesome.

Comment (costinm) [reply pilot/docker/Dockerfile.proxy_debug]: Istionightly is the repo configured for the CircleCI - we want to built it nightly (and automatically).
I don't want CircleCI to have permissions for istio/ org.

For release: if the image is built first it'll use the local version instead of downloading.

File: pilot/docker/Dockerfile.proxy_debug (hunk 2 above)
Comment (costinm) [reply pilot/docker/Dockerfile.proxy_debug]: In the base image.

Comment (costinm) [reply pilot/docker/Dockerfile.proxy_debug]: Ah, sorry missed the add. See docker/Dockerfile.xenial_debug

File: pilot/docker/Dockerfile.proxy_debug (hunk 1 above)
Comment (costinm) [reply pilot/docker/Dockerfile.proxy_debug]: It takes very long to re-build the base image - I want it to be built by the nightly job.

Also I want to be able to optionally use the debug base image for the pilot test app ( and possibly pilot, mixer ).

File: docker/Dockerfile.xenial_debug [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      sudo &&  apt-get upgrade -y && \
+    rm -rf /var/lib/apt/lists/*
+
+# Sudoers used to allow tcpdump and other debug utilities.
+RUN mkdir /var/lib/istio && useradd -m --uid 1337 istio-proxy && \
+    echo "istio-proxy ALL=NOPASSWD: ALL" >> /etc/sudoers && \
+    chown -R istio-proxy /var/lib/istio
+
+
+# TODO: add delve
```
Comment (ldemailly) [inline docker/Dockerfile.xenial_debug]: missing newline

which target builds it ? will it get built by the release process ?

File: pilot/docker/Dockerfile.proxy_debug (hunk 1 above)
Comment (nmittler) [reply pilot/docker/Dockerfile.proxy_debug]: ok, that seems reasonable.  To be clear: there is no istio-specific stuff going on in the base image, correct?

Comment (ldemailly) [reply pilot/docker/Dockerfile.proxy_debug]: will the release process build it by default (how/without change? or which change)? why don't we use the (official) dailies or the latest version and rebuild it as part of the actual release process

cc @hklai

Comment (costinm) [reply pilot/docker/Dockerfile.proxy_debug]: Release script needs to call 'docker.basedebug' first. Will check with the release team to make sure this happens.

Comment (costinm) [reply pilot/docker/Dockerfile.proxy_debug]: There is some istio-specific stuff right now ( I can move it back ) - there is an istio-proxy user created. I don't think it hurts in any way to have the user - but it'll help for some tests and 
use cases. 

The base image is just a refactoring of what we had - to avoid 'apt-get update /install' in each
build and to allow some reuse.

Comment (nmittler) [reply pilot/docker/Dockerfile.proxy_debug]: @costinm Yeah, let's move it back.  Having istio-specific stuff stuck in a base image that isn't part of the local build is troublesome for development.

Comment (ldemailly) [reply pilot/docker/Dockerfile.proxy_debug]: I think we should treat the release as an (important) api and not keep making changes that require downstream processes to change to adapt to circleci or some dev specific workflows - those should go into special targets

Comment (costinm) [reply pilot/docker/Dockerfile.proxy_debug]: Done.

Comment (costinm) [reply pilot/docker/Dockerfile.proxy_debug]: Release is an important step, and I agree we should test it and including in the build process.

I don't mind someone adding a 'make release' target, and moving the gcloud-specific commands to a separate upload step.

File: tests/e2e/tests/pilot/testdata/pilot.yaml.tmpl [hunk 4]
```diff
@@ -42,6 +42,8 @@ spec:
   type: LoadBalancer
 {{end}}
   ports:
+  - port: 8080
```
Comment (ZackButcher) [inline tests/e2e/tests/pilot/testdata/pilot.yaml.tmpl]: How much is it going to be to use the helm charts here rather than these testdata template files? (Does it make sense to do in this PR?)

File: tests/istio.mk [hunk 5]
```diff
@@ -83,35 +83,82 @@ e2e_all: | $(JUNIT_REPORT)
 	$(MAKE) e2e_simple e2e_mixer e2e_bookinfo \
 	|& tee >($(JUNIT_REPORT) > $(JUNIT_E2E_XML))
 
+# Run the e2e tests, with auth enabled. A separate target is used for non-auth.
 e2e_pilot: istioctl generate_yaml
 	go test -v -timeout 20m ./tests/e2e/tests/pilot ${TESTOPTS} -hub ${HUB} -tag ${TAG}
 
+e2e_pilot_noauth: istioctl generate_yaml
+	go test -v -timeout 20m ./tests/e2e/tests/pilot ${E2E_ARGS} -hub ${HUB} -tag ${TAG} --skip-cleanup -mixer=true -auth=disable -use-sidecar-injector=false
```
Comment (ZackButcher) [inline tests/istio.mk]: please remove `--skip-cleanup` from target

why no sidecar injector?

File: tests/istio.mk [hunk 6]
```diff
@@ -83,35 +83,82 @@ e2e_all: | $(JUNIT_REPORT)
...
+        #   |& tee >($(JUNIT_REPORT) > ${OUT_DIR}/tests/test-report-auth-pilot.xml)
+
+
+# Target for running e2e pilot in a minikube env. Used by CI
+test/minikube/noauth/e2e_pilot: istioctl
+	mkdir -p ${OUT_DIR}/logs
+	mkdir -p ${OUT_DIR}/tests
+	# istio-system and pilot system are not compatible. Once we merge the setup it should work.
+	kubectl create ns pilot-system-test || true
+	kubectl create ns pilot-test || true
```
Comment (ZackButcher) [inline tests/istio.mk]: we should refactor out this common test setup into its own target that these invocations can share (not blocking on this PR, just a longer term improvement we should make).

File: tests/istio.mk [hunk 7]
```diff
@@ -83,35 +83,82 @@ e2e_all: | $(JUNIT_REPORT)
...
+		--use-sidecar-injector=false \
+		--auth_enable=false \
+		-v1alpha3=true -v1alpha1=false \
+		--core-files-dir=${OUT_DIR}/logs \
+        	--ns pilot-system-test \
+        	-n pilot-test \
            ${TESTOPTS}
+           # |& tee >($(JUNIT_REPORT) > ${OUT_DIR}/tests/test-report-noauth-pilot.xml)
 
 # Target for running e2e pilot in a minikube env. Used by CI
```
Comment (ZackButcher) [inline tests/istio.mk]: Is this still accurate?

File: tests/e2e/tests/pilot/testdata/pilot.yaml.tmpl (hunk 4 above)
Comment (costinm) [reply tests/e2e/tests/pilot/testdata/pilot.yaml.tmpl]: Obviously not in this PR... Mandar is working on it, and it's a very different problem.

File: tests/istio.mk (hunk 5 above)
Comment (costinm) [reply tests/istio.mk]: It is quite important:
- in minikube, the image can be wiped
- in dev mode - you may want to reuse or test or run individual checks manually after the test is run.

That's what (and how) we've been testing in circle - I just moved them here so it's easier to reproduce, and they can be used by developers. 

I found that every time the test fails - or even if it works - it's good to have the environment for
additional work. There is already a target that doesn't do this.

File: tests/istio.mk (hunk 6 above)
Comment (costinm) [reply tests/istio.mk]: Yes, creating a stable test configuration first - and running the test after without Sleep is quite good, but let's not do everything in one PR. Big +1 if you want to do it.

File: tests/istio.mk (hunk 7 above)
Comment (costinm) [reply tests/istio.mk]: Well, I saw a green build - even if the tests were failing. 

I'll keep working on it - but don't want to hold this PR until I figure it out, and is not critical.
I can remove the commented out line if it hurts anyone.

File: tests/istio.mk [hunk 8]
```diff
@@ -83,35 +83,82 @@ e2e_all: | $(JUNIT_REPORT)
...
 
+e2e_pilot_noauth: istioctl generate_yaml
+	go test -v -timeout 20m ./tests/e2e/tests/pilot ${E2E_ARGS} -hub ${HUB} -tag ${TAG} --skip-cleanup -mixer=true -auth=disable -use-sidecar-injector=false
+
+test/minikube/auth/e2e_simple:
+	go test -v -timeout 20m ./tests/e2e/tests/simple -args --auth_enable=true \
+	  --skip_cleanup  -use_local_cluster -cluster_wide -test.v \
+	  ${E2E_ARGS} ${EXTRA_E2E_ARGS} \
+           ${TESTOPTS}
+           #|& tee >($(JUNIT_REPORT) > ${OUT_DIR}/tests/test-report-auth-simple.xml)
```
Comment (chxchx) [inline tests/istio.mk]: This comment applies to all the commented code to generate junit xml.

If you don't need Junit reports just yet then remove the comment and add it back in another PR when you need them. If you do need it, then you must have `set -o pipefail;` in the same line since
* make executes each line in a subshell
* not setting pipefail will let go-junit-report mask the test failures since go-junit-report alomst always returns zero

And you need to have a order-only dependency on `$(JUNIT_REPORT)` to make sure it is installed for everyone, CI or developers. The make target `e2e_all` is a good example to follow.

Comment (costinm) [reply tests/istio.mk]: Yes, we do need junit reports - but I don't want to make it too complicated. 
I'll change the circleci config instead to run it at the end, found a cleaner way.

File: pilot/docker/Dockerfile.proxy_debug [hunk 9]
```diff
@@ -12,7 +12,7 @@ COPY envoy_mixer_auth.json /etc/istio/proxy/envoy_mixer_auth.json
 COPY envoy_bootstrap_tmpl.json /var/lib/istio/envoy/envoy_bootstrap_tmpl.json
 
 # Sudoers used to allow tcpdump and other debug utilities.
-RUN mkdir /var/lib/istio && useradd -m --uid 1337 istio-proxy && \
+RUN useradd -m --uid 1337 istio-proxy && \
```
Comment (ldemailly) [inline pilot/docker/Dockerfile.proxy_debug]: what is creating /var/lib/istio ?

Comment (costinm) [reply pilot/docker/Dockerfile.proxy_debug]: Removed it, not sure ( it was there before ).

**Comment by costinm** [comment]
Ping ? Can anyone approve it ?

**Comment by mattdelco** [comment]
As food for thought, or as an alternative for some of the goals of this PR (either instead of, or subsequent update to parts of this PR):

https://github.com/istio/istio/compare/master...mattdelco:master20

It basically covers building all of the docker images in a unique directory.  Quicker rebuilds would be handling by passing SKIP_REBUILD_CHECK=1 to make (mostly to opt-out of giving golang a chance to rebuild) rather than setting up different targets with different semantics.  The behavior is easier to rationalize when the dependency lists cite concrete files/paths rather than phony ones (which tend to always get rebuilt by make).

The other part of that PR also gets rid of bin/init.sh (by porting it into the Makefile, and also leveraging fortio to take the place of curl/wget).

The change is basically complete other than me having to repeatedly catch up with rebasing and incorporating other changes.  Given the focus on tests this week I didn't expect so much activity to be going on with the Makefiles this week.

**Comment by ldemailly** [comment]
@mattdelco https://github.com/istio/istio/compare/master...mattdelco:master20 is nice, thanks ! make a PR ?
(or maybe after this one...)

File: pilot/docker/Dockerfile.proxy_debug (hunk 1 above)
Comment (ldemailly) [reply pilot/docker/Dockerfile.proxy_debug]: any existing target you know the release already calls that you can add `docker.basedebug` to ? @chxchx do you know ?

File: docker/Dockerfile.xenial_debug [hunk 10]
```diff
@@ -0,0 +1,24 @@
+FROM ubuntu:xenial
+# Base image for debug builds.
+# Built manually uploaded as "istionightly/base_debug"
+
+# Do not add more stuff to this list that isn't small or critically useful.
+# If you occasionally need something on the container do
+# sudo apt-get update && apt-get whichever
+RUN apt-get update && \
+    apt-get install --no-install-recommends -y \
+      curl \
```
Comment (ldemailly) [inline docker/Dockerfile.xenial_debug]: can we add fortio :-)

File: tests/e2e/tests/pilot/testdata/pilot.yaml.tmpl (hunk 4 above)
Comment (costinm) [reply tests/e2e/tests/pilot/testdata/pilot.yaml.tmpl]: It almost never makes sense to ask for more features in a PR ( sometimes it makes sense to ask to split the PR ).

Comment (costinm) [reply tests/e2e/tests/pilot/testdata/pilot.yaml.tmpl]: (it's not trivial - the pilot init is very different and incompatible )

**Comment by costinm** [comment]
Matt: please don't submit that PR ! I am ok with the extra 'don't check' variable - but moving the download into Makefile doesn't make sense, and absolutely -1 on more 'for' and macros. Enough pain unwinding the docker macros. Yes, make can be used in crazy ways - but we're not trying to turn it into another bazel.

File: docker/Dockerfile.xenial_debug (hunk 10 above)
Comment (costinm) [reply docker/Dockerfile.xenial_debug]: I know you like fortio  - but please don't add more stuff here. There is a docker image for it.

File: .circleci/config.yml [hunk 11]
```diff
@@ -52,16 +52,28 @@ jobs:
...
               time make build test-bins
             fi
-            make docker.tag generate_yaml
+            make docker.all generate_yaml
       - run: bin/testEnvRootMinikube.sh wait
       - run: docker images
-      - run: PATH=$GOPATH/bin:$PATH make e2e_simple_run E2E_ARGS="--skip_cleanup -use_local_cluster -cluster_wide -test.v"
+      - run: make test/minikube/noauth/e2e_simple
+      - run:
+          name: dumpsys
```
Comment (kyessenov) [inline .circleci/config.yml]: Use YAML references to avoid repetition?

File: .circleci/config.yml [hunk 12]
```diff
@@ -491,12 +516,14 @@ jobs:
...
           docker_layer_caching: true
+      - run: make istio-archive
       - run:
           command: |
             if [ ! -z "${DOCKER_USER}" ] ; then
               echo "Pushing docker images"
               docker login -u ${DOCKER_USER} -p ${DOCKER_PASS}
-              make docker push
+              # The xenail debug image is built first, will be used in the next steps
+              make docker.push.basedebug docker.all push
```
Comment (kyessenov) [inline .circleci/config.yml]: docker.basedebug.push to be consistent with docker.pilot and other docker.<image> rules?

File: .circleci/config.yml [hunk 13]
```diff
@@ -491,12 +516,14 @@ jobs:
...
       - setup_remote_docker:
           docker_layer_caching: true
+      - run: make istio-archive
       - run:
           command: |
             if [ ! -z "${DOCKER_USER}" ] ; then
               echo "Pushing docker images"
               docker login -u ${DOCKER_USER} -p ${DOCKER_PASS}
-              make docker push
+              # The xenail debug image is built first, will be used in the next steps
```
Comment (kyessenov) [inline .circleci/config.yml]: _xenial_

File: docker/Dockerfile.xenial_debug (hunk 10 above)
Comment (kyessenov) [inline docker/Dockerfile.xenial_debug]: How about https://github.com/envoyproxy/envoy-tools/tree/master/envoy-curses?

File: pilot/docker/Dockerfile.proxy_debug [hunk 14]
```diff
@@ -1,34 +1,14 @@
...
-      netcat \
-      tcpdump \
-      net-tools \
-      libc6-dbg gdb \
-      elvis-tiny \
-      lsof \
-      busybox \
-      sudo && \
-    rm -rf /var/lib/apt/lists/*
+FROM istionightly/base_debug
```
Comment (kyessenov) [inline pilot/docker/Dockerfile.proxy_debug]: Shouldn't this be a tagged image for reproducibility? How are we going to rebuild this image in a month time?

File: tests/e2e/tests/pilot/testdata/config.yaml.tmpl [hunk 15]
```diff
@@ -11,12 +11,12 @@ data:
       drainDuration:          2s
       parentShutdownDuration: 3s
 {{if eq .Auth.String "NONE"}}
-      discoveryAddress:       istio-pilot.{{.IstioNamespace}}:15007
+      discoveryAddress:       istio-pilot.{{.IstioNamespace}}:8080
```
Comment (kyessenov) [inline tests/e2e/tests/pilot/testdata/config.yaml.tmpl]: Why 8080? We don't want observability anymore for pilot? This is not a build and test improvement, it changes how the system behaves.

File: tests/e2e/tests/pilot/testdata/config.yaml.tmpl [hunk 16]
```diff
@@ -11,12 +11,12 @@ data:
...
 {{else}}
       discoveryAddress:       istio-pilot.{{.IstioNamespace}}:15005
 {{end}}
-      discoveryRefreshDelay:  1s
-      connectTimeout:         1s
+      discoveryRefreshDelay:  2s
```
Comment (kyessenov) [inline tests/e2e/tests/pilot/testdata/config.yaml.tmpl]: Why is this necessary now?

Comment (kyessenov) [reply tests/e2e/tests/pilot/testdata/config.yaml.tmpl]: I mean this is going to make every test to flake at the first iteration. There is a fixed 3s delay between rule application. You need to increase that delay if we you want to increase refresh delay, or you're just racing them.

File: tests/e2e/tests/pilot/testdata/config.yaml.tmpl [hunk 17]
```diff
@@ -27,8 +27,8 @@ data:
     mixerReportServer:        istio-mixer.{{.IstioNamespace}}:15004
 {{end}}
     authPolicy:               {{ .Auth }}
-    connectTimeout:           1s
-    rdsRefreshDelay:          1s
+    connectTimeout:           5s
+    rdsRefreshDelay:          4s
```
Comment (kyessenov) [inline tests/e2e/tests/pilot/testdata/config.yaml.tmpl]: Why? This is going to fail every single config test at the first retry.

**Comment by kyessenov** [comment]
Costin, please explain why you changed ports and mesh configs without adjusting the tests accordingly.

File: docker/Dockerfile.xenial_debug (hunk 10 above)
Comment (ldemailly) [reply docker/Dockerfile.xenial_debug]: it’s 2mb and is useful (to me and orlthers for debugging)

**Comment by kyessenov** [comment]
Sorry, clicked wrong button to open/close... Apologies.

**Comment by ldemailly** [comment]
> And absolutely don't replace 'curl' with fortio in scripts. Inventing our own downloader ? C'mon... I understand using it for load tests - but it is not a dependency of istio.

you miss the context earlier, we have ugly code trying to switch between curl and wget and check if curl has tls support, matt pointed out we have a dependency on go already and wanted to write a short wrapper around go’s http client, I pointed out fortio already has that exact same code and is already in vendor and used for tests so not a new dependency

**Comment by costinm** [comment]
Happy to remove the ugly code dealing with wget - we require a working curl, that's not unreasonable. 
It doesn't mean because some user can't get a good 'curl' we need to reimplement it or use fortio as curl.

File: tests/e2e/tests/pilot/testdata/config.yaml.tmpl (hunk 17 above)
Comment (costinm) [reply tests/e2e/tests/pilot/testdata/config.yaml.tmpl]: The tests seem to be passing with 4 sec - and yes, they may fail on the first retry, as they should. 

We have to increase the refresh - since it cant' scale, and tests should reflect this.

File: tests/e2e/tests/pilot/testdata/config.yaml.tmpl (hunk 16 above)
Comment (costinm) [reply tests/e2e/tests/pilot/testdata/config.yaml.tmpl]: To test with realistic values - production can't really work with 1 sec refresh.

**Comment by ldemailly** [comment]
you shouldn't merge when there are outstanding questions

File: tests/e2e/tests/pilot/testdata/config.yaml.tmpl (hunk 15 above)
Comment (costinm) [reply tests/e2e/tests/pilot/testdata/config.yaml.tmpl]: The 15007 feature was half done and not working with larger number of pods. Testing with what we'll use.

File: pilot/docker/Dockerfile.proxy_debug (hunk 14 above)
Comment (costinm) [reply pilot/docker/Dockerfile.proxy_debug]: We can tag it - in a separate PR. This is not a regression - before an apt-get update was also made,
I just moved that to a different image.

This doesn't have to be pushed or downloaded - part of the build you can rebuild.
